File: cgi-bin/bulk_pilot_import.pl
#!/usr/bin/perl -w
#
# Reads in a CSV membership list
#
#

require DBD::mysql;
use Data::Dumper;

# Add currect bin directory to @INC
use File::Basename;
use lib '/home/untps52y/perl5/lib/perl5';
use lib dirname (__FILE__) . '/';
use TrackLib qw(:all);

#
# Database handles
#


#
# Extract 'fix' data into a nice record
#
sub read_membership
{
    my ($dbh,$f) = @_;

    my @field;
    my $row;
    my $nat;

    print "reading: $f\n";
    open(FD, "$f") or die "can't open $f: $!";

    while (<FD>)
    {
        $row = $_;

        print "row=$row\n";

        @field = split /,/, $row;
        if ($field[1] eq '')
        {
            next;
        }
        print 'add name: ', $field[1], ' ', $field[2], "\n";
        
        # Get Country ID
        $nat = $field[3];
        $sth = $dbh->prepare("	SELECT 
									C.natID AS Code 
								FROM 
									tblCountryCodes C 
								WHERE 
									C.natIso3 = '$nat'");
        $sth->execute();
        if  ($ref = $sth->fetchrow_hashref())
		{
			$nat = $ref->{'Code'};
		}
		else
		{
			# Set Italy
			$nat = 380;
		}
        


        # should be an insert/update ..
        insertup('tblExtPilot', 'pilPk', "pilLastName='" . $field[2] . 
                "' and pilFAI='" . $field[0] . "'",
            {
                'pilFirstName' => $field[1],
                'pilLastName' => $field[2],
                'pilFAI' => $field[0],
                'pilSex' => $field[4],
                'pilNat' => $nat
            });
        #$dbh->do("INSERT INTO tblPilot (pilFirstName,pilLastName,pilFAI,pilSex) VALUES (?,?,?,?)", undef, $field[1],$field[0],$field[2],'M');
    }

}

#
# Main program here ..
#

my $flight;
my $allflights;
my $traPk;
my $totlen = 0;
my $coords;
my $numc;
my $numb;
my $score;
my $dbh;

$dbh = db_connect();

print 'Arg: ' . $ARGV[0];

if (scalar @ARGV < 1)
{
    print "bulk_pilot_import.pl <membershiplist>\n";
    exit(1);
}

# Read the csv members
$members = read_membership($dbh, $ARGV[0]);

File: cgi-bin/bulk_pilot_import.py
"""
Reads in a CSV membership list
Use: python3 bulk_pilot_import.py [csv file name]

Antonio Golfari - 2018
"""

# Use your utility module.
from compUtils  import *
from logger     import Logger
import csv
from pathlib    import Path


def read_membership(file):

    """Read CSV File"""
    with open(file, encoding='utf-8') as csv_file:
        csv_reader = csv.reader(csv_file, delimiter=',')
        for row in csv_reader:
            pilFAI = row[0]
            pilFName = row[1]
            pilLName = row[2]
            pilSex = row[4]
            nat = row[3]

            """Get Nat code"""
            pilNat = get_nat_code(nat)
            if pilNat is None:
                """defaults to Italy"""
                pilNat = 380
                print(f"Pilot {pilFName} {pilLName} has a unrecognized country code, defaulting to ITA \n")

            with Database() as db:
                """Check if pilot exists in pilots main table"""
                query = """SELECT
                                `pilPK` AS `ID`
                            FROM `PilotView`
                            WHERE
                                (`pilLastName` LIKE '%%{0}%%' AND `pilFirstName` LIKE '%%{1}%%')
                            OR (`pilLastName` LIKE '%%{0}%%' AND `pilFAI` = {2})
                            LIMIT 1"""
                params = [pilLName, pilFName, pilFAI]

                try:
                    """Pilot exists already"""
                    # print ("Query: {} \n Rows: {} \n".format(query, db.rows(query)))
                    pil = 0 + db.fetchone(query, params)['ID']
                    print(f"pilot {pilLName} {pilFName} exists in pilots list: id {pil} \n")
                except:
                    """Check if FAI exists"""
                    query = """SELECT
                                    `pilPK` AS `ID`, `pilLastName` AS `Name`
                                FROM `PilotView`
                                WHERE `pilFAI` = %s LIMIT 1"""
                    try:
                        """Pilot exists already"""
                        row = db.fetchone(query, [pilFAI])
                        print(f"FAI n. {pilFAI} already associated to pilot {row['Name']}: id {row['ID']} \n")
                    except:
                        """Insert new pilot"""
                        query = """ INSERT INTO `tblExtPilot`
                                        (`pilFirstName`,
                                        `pilLastName`,
                                        `pilNat`,
                                        `pilSex`,
                                        `pilFAI`)
                                    VALUES
                                        (%s,%s,%s,%s,%s)"""
                        params = [pilFName, pilLName, pilNat, pilSex, pilFAI]

                        """Normal Mode"""
                        pil_id = db.execute(query, params)
                        print(f"PILOT {pilLName} {pilFName} ADDED with id {pil_id}\n")

def main(args):
    """Main module"""
    my_file = Path(str(args[0]))

    '''create logging and disable output'''
    Logger('ON', 'bulk_pilots_import.txt')

    """Check if file exists"""
    if not my_file.is_file():
        print (f"Reading error: {str(args[0])} does not exist")
        ''' restore stdout function '''
        Logger('OFF')
        print(0)
        exit()

    """Read CSV File"""
    read_membership(file)

    ''' restore stdout function '''
    Logger('OFF')
    print(1)

if __name__ == "__main__":
    import sys

    '''check parameter is good'''
    if not (sys.argv[1]

            and len(sys.argv) == 3):
        print("number of arguments != 2 and/or task_id not a number")
        print("Use: python3 bulk_igc_reader.py <taskPk> <zipfile>")
        exit()

    main(sys.argv[1:])
